src/Splat.Microsoft.Extensions.DependencyInjection/MicrosoftDependencyResolver.cs

```diff
@@ -166,22 +164,21 @@ public virtual void UnregisterCurrent(Type? serviceType, string? contract = null
         {
             if (contract is null || string.IsNullOrWhiteSpace(contract))
             {
-                var sd = _serviceCollection?.LastOrDefault(s => s.ServiceType == serviceType);
+                var sd = _serviceCollection?.LastOrDefault(s => !s.IsKeyedService && s.ServiceType == serviceType);
                 if (sd is not null)
                 {
                     _serviceCollection?.Remove(sd);
                 }
             }
             else
             {
-                var dic = GetContractDictionary(serviceType, false);
-                if (dic is not null)
+                var sd = _serviceCollection?.LastOrDefault(sd => sd.IsKeyedService
+                                                            && sd.ServiceKey is string serviceKey
+                                                            && serviceKey == contract
+                                                            && sd.ServiceType == serviceType);
```

Comment:
putting "&& sd.ServiceType == serviceType" above the "&& sd.ServiceKey is string serviceKey" will short cicruit before the cast and string equality check.